testbed/core/views.py

```diff
@@ -72,3 +82,200 @@ def index(request):
+        'state': state
+    }
+    
+    # Validate the state parameter to prevent CSRF attacks
+    if not state:
+        logger.warning("OAuth callback received with no state parameter")
+        context['error'] = 'invalid_state'
+        context['error_description'] = 'No state parameter provided'
+    else:
+        is_valid_state = validate_state_from_session(request, state)
+        if not is_valid_state:
+            logger.warning("OAuth callback received with invalid state parameter")
+            context['error'] = 'invalid_state'
+            context['error_description'] = 'Invalid state parameter'
+    
+    # Log successful authorization
+    if code and not error and not context.get('error'):
+        logger.info("Successfully received authorization code in callback")
+    
+    return render(request, 'oauth_callback.html', context)
+
+@login_required
+def test_authorization_view(request):
+    """
+    Test view to simulate initiating an OAuth authorization flow.
+    This will redirect to the authorization endpoint with appropriate parameters.
+    """
+    # Get the user's OAuth application (this is normally used by the destination service)
+    application = get_user_application(request.user)
+    
+    # Use our own callback URL for the test flow
+    scheme = request.scheme
+    host = request.get_host()
+    redirect_uri = f"{scheme}://{host}/callback"
+    
+    # Update the application's redirect URIs if needed
+    if redirect_uri not in application.redirect_uris:
+        if application.redirect_uris:
+            application.redirect_uris = f"{application.redirect_uris} {redirect_uri}"
+        else:
+            application.redirect_uris = redirect_uri
+        application.save()
+        logger.info(f"Added {redirect_uri} to redirect URIs for application {application.client_id}")
+    
+    # Generate a secure random state parameter and store it in the session
+    state = generate_secure_state()
+    store_state_in_session(request, state)
+    
+    # Build the authorization URL
+    params = {
+        'client_id': application.client_id,
+        'response_type': 'code',
+        'scope': 'activitypub_account_portability',
+        'redirect_uri': redirect_uri,
+        'state': state
+    }
+    
+    # Convert params to URL query string
+    query_string = '&'.join([f"{key}={value}" for key, value in params.items()])
+    auth_url = f"{reverse('oauth2_provider:authorize')}?{query_string}"
+    
+    return redirect(auth_url)
+
+@login_required
+def test_error_view(request):
+    """
+    Test view to simulate an OAuth error.
+    This renders the error template with a sample error.
+    """
+    error = {
+        'error': 'invalid_request',
+        'description': 'This is a test error to preview the error template.',
+        'uri': None
+    }
+    
+    return render(request, 'oauth2_provider/error.html', {'error': error})
+
+
+# Token Exchange View
+@login_required
+def test_token_exchange_view(request):
```

Comment:
When we iterate on this, is token exchange triggered automatically on getting an authorization code?  When this is integrated into a working setup to do account portability, would the callback view that we think gets the authorization code just move straight to triggering the token exchange code (without having a URL that some browser has to trigger?)